app/plugin_key_value_store.go

```diff
@@ -57,6 +65,17 @@ func (a *App) GetPluginKey(pluginId string, key string) ([]byte, *model.AppError
 
 	kv := result.Data.(*model.PluginKeyValue)
 
+	// If we are here that means we are using old hashed key. Remove it and insert the new key without hashing it
```

Comment:
I'm inclined not to write to the key value storage on a get operation.

I assume the thinking here is to slowly migrate the keying strategy, but completing that assumes 1) all input keys get fetched in a timely fashion by "natural processes", and 2) we have some way of knowing that we're done converting keys. Since we can't guarantee the former (who knows how plugins use keys), and we can't detect the latter (no way to provably distinguish a hashed key from an unhashed one unless we add a column), I don't think there's a future where we can remove this code. In that case, is there any point in the migration?

Reply:
Considering plugins are still in beta, and we would update the key format in gets and sets, it would be reasonable to give a migration period of 2-3 releases and then remove the fallback code. Adding a temporary column wouldn't give us any advantage, considering even if we _know_ there are still some outdated keys in the DB, we cannot find out what's their key. Supporting hashed keys forever doesn't sound like a good approach for me.

What you think @jwilander @lieut-data ?

Reply:
What's the downside of keeping the code around forever? Having to maintain it and any misses on the regular key cause another get for the hashed key? That's not terrible. My worry about trying to migrate as Jesse says by "natural processes" is that in 2-3 releases when we make the breaking change we'll 1) be stable for plugins and 2) it could quietly cause issues for some plugins without any clear path for them to notice or fix the problem.

So if we're going to make a breaking change I think we should do it now and if we don't want to do that I think we need to keep the code around indefinitely unless it becomes a real problem for whatever reason

Reply:
No, as you've said yourself it's neither a high performance impact, or is the code really hard to read that way. I was mostly questioning whether it's worth to break the current API (and essentially force a wipe of the KV Table with the next migration, since it's worthless data) to clean up the schema and the old hashed key format, or we are fine with keeping the code for future.
Personally not having a particular strong feeling about either option.

Reply:
Forcing a wipe seems unusually painful -- e.g., this would force all the GitHub plugin users to re-sign in, for example.

With respect to the original question, another concern is that if you list the keys, then get a key, then list the keys again, the set will change, perhaps unexpectedly. It seems reasonable to do this on an `UPDATE` (changing the keys), but surprising otherwise.

Reply:
Make sense. Do we have a consensus to not update records on a get?

Reply:
I'm for not updating the keys

Reply:
One upside of updating the keys is that `KVList()` becomes more useful. It we never update the keys plugins will never be able to list "old" stored keys.

Reply:
`KVList` /should/ be just as useful in terms of listing the keys that can be fetched, but I agree that we lose an opportunity to "convert" back to the original key, until the plugin decides to write it again. I still think the "principle of no surprise" is best applied here, since we can't guarantee ever recovering the original keys at all, and we do still at least try to update them on set.

Reply:
That makes sense, so the consensus appears to be to _not_ update the records on a get @gupsho - Sorry for the discussing as part of your PR.

Reply:
Let me update it.